docs/our-approach-to-data.md

```diff
@@ -271,3 +272,137 @@ By now, you're hopefully convinced that a global application state can enable us
 
 Files related to user-interface state can be found in the [`client/state/ui` directory](../client/state/ui).
+
+### Data Persistence ( [#2754](https://github.com/Automattic/wp-calypso/pull/2754) )
+
+Persisting our Redux state to browser storage (localStorage/indexedDB) allows us to avoid completely rebuilding the 
+Redux tree from scratch on each page load.
+
+At a high level, implementing this is straightforward. We subscribe to any Redux store changes, and on change we update 
+our browser storage with the new state of the Redux tree. On page load, if we detect stored state in browser storage during 
+our initial render, we create our Redux store with that persisted initial state.
+ 
+However we quickly run into the following problems:
+
+#### Subtrees may contain class instances
+
+Some subtrees contain class instances. In some cases this is expected, as some branches have chosen to use
+Immutable.js for performance reasons. However, attempting to serialize class instances will throw errors while saving 
+to our browser storage.
+
+#### Data shapes change over time ( [#3101](https://github.com/Automattic/wp-calypso/pull/3101) )
+
+As time passes, the shape of our data will change very drastically in our Redux store and in each subtree. If we now
+persist state, we run into the issue of our data shapes no longer matching. 
+
+As a developer, this case is extremely easy to hit. If Redux persistence is enabled and we are running master, first allow 
+state to be persisted to the browser and then switch to another branch that has some minor refactors for an 
+existing sub-tree. What happens when a selector reaches for a data property that doesn't exist, or has been renamed? 
+Errors, of course!
+
+A normal user can hit this case too, by visiting our website and returning two weeks later.
+
+How can we tell that our persisted data is good to use as initial state?
+
+#### Serialize and Deserialize
+
+To work around these issues, we create two special actions: `SERIALIZE` and `DESERIALIZE`. These actions are not 
+dispatched, but are instead used with the reducer directly to prepare state to be serialized to browser storage, 
+and for deserializing persisted state to an acceptable initialState for the Redux store.
+
+
+```javascript
+reducer( reduxStore.getState(), { type: 'SERIALIZE' } )
+```
+and
+
+```javascript
+reducer( browserState, { type: 'DESERIALIZE' } )
+```
+
+The key idea is that in the reducer `SERIALIZE` should return a plain JS object. In a subtree that uses Immutable.js
```

Comment:
Should we explain why it's important to return a JavaScript object?

```text
Because browser storage is only capable of storing simple JavaScript objects, the purpose of the `SERIALIZE` action type reducer handler is to return a plain object representation.
```